File: flutter_app/lib/models/book_entity.dart
// Mirror of ios/EpubToMp3/EpubToMp3/Models/BookEntity.swift @ b7c962a
//
// One book in the user's local library. Library is the hero UI on the
// Apple side; we mirror the data shape + status logic so the Flutter
// LibraryStore can persist the same JSON layout under
// `library.books.v1` in SharedPreferences.
//
// Differences vs Swift:
//   - `bookmark` (security-scoped bookmark Data) has no Flutter
//     equivalent. We persist the absolute file path instead. macOS /
//     iOS aren't supported targets for the Flutter app (see project
//     scope memory), so the sandbox bookmark concept is moot here.
//   - `coverPNG` -> base64-encoded String, since SharedPreferences
//     cannot store raw bytes directly.

import 'dart:convert';

enum LibraryStatus { textOnly, caching, offlineReady }

extension LibraryStatusX on LibraryStatus {
  String get rawValue => switch (this) {
        LibraryStatus.textOnly => 'textOnly',
        LibraryStatus.caching => 'caching',
        LibraryStatus.offlineReady => 'offlineReady',
      };
}

class BookEntity {
  /// Stable id derived from the SHA-256 of the EPUB file contents.
  final String id;
  String title;
  String? author;

  /// Absolute path on disk. Flutter does not have iOS-style
  /// security-scoped bookmarks; if the file is gone the UI re-prompts.
  String filePath;

  /// Display-only filename. Use [filePath] for I/O.
  final String displayFilename;

  final DateTime addedAt;
  DateTime? lastOpenedAt;

  int? lastChapterIndex;
  double? lastPositionSeconds;

  /// PNG/JPEG bytes for cover art, base64-encoded for prefs storage.
  String? coverBase64;

  /// Most recent conversion jobId, to reattach to backend audio assets.
  String? lastJobId;

  /// User opted in to caching audio offline.
  bool cachedOffline;

  BookEntity({
    required this.id,
    required this.title,
    this.author,
    required this.filePath,
    required this.displayFilename,
    required this.addedAt,
    this.lastOpenedAt,
    this.lastChapterIndex,
    this.lastPositionSeconds,
    this.coverBase64,
    this.lastJobId,
    this.cachedOffline = false,
  });

  LibraryStatus get status {
    if (cachedOffline) return LibraryStatus.offlineReady;
    if (lastJobId != null) return LibraryStatus.caching;
    return LibraryStatus.textOnly;
  }

  String get resolvedTitle {
    final t = title.trim();
    return t.isEmpty ? displayFilename : t;
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        if (author != null) 'author': author,
        'filePath': filePath,
        'displayFilename': displayFilename,
        'addedAt': addedAt.toIso8601String(),
        if (lastOpenedAt != null)
          'lastOpenedAt': lastOpenedAt!.toIso8601String(),
        if (lastChapterIndex != null) 'lastChapterIndex': lastChapterIndex,
        if (lastPositionSeconds != null)
          'lastPositionSeconds': lastPositionSeconds,
        if (coverBase64 != null) 'coverBase64': coverBase64,
        if (lastJobId != null) 'lastJobId': lastJobId,
        'cachedOffline': cachedOffline,
      };

  factory BookEntity.fromJson(Map<String, dynamic> json) => BookEntity(
        id: json['id'] as String,
        title: json['title'] as String? ?? '',
        author: json['author'] as String?,
        filePath: json['filePath'] as String? ?? '',
        displayFilename: json['displayFilename'] as String? ?? '',
        addedAt: DateTime.tryParse(json['addedAt'] as String? ?? '') ??
            DateTime.fromMillisecondsSinceEpoch(0),
        lastOpenedAt: json['lastOpenedAt'] is String
            ? DateTime.tryParse(json['lastOpenedAt'] as String)
            : null,
        lastChapterIndex: (json['lastChapterIndex'] as num?)?.toInt(),
        lastPositionSeconds:
            (json['lastPositionSeconds'] as num?)?.toDouble(),
        coverBase64: json['coverBase64'] as String?,
        lastJobId: json['lastJobId'] as String?,
        cachedOffline: json['cachedOffline'] as bool? ?? false,
      );

  String encode() => jsonEncode(toJson());

  static BookEntity decode(String s) =>
      BookEntity.fromJson(jsonDecode(s) as Map<String, dynamic>);
}

File: ios/EpubToMp3/EpubToMp3/Features/Library/Models/BookEntity.swift
import Foundation

/// Discriminator between the two natively supported book formats. The
/// Library opens both with the same picker / drop / share surface,
/// then routes to the right reader at open time.
enum BookFileType: String, Codable, Hashable, CaseIterable {
    case epub
    case pdf

    /// Best-effort classification by extension. Anything we cannot
    /// classify falls back to `.epub` because the EPUB reader is
    /// tolerant of non-zip inputs (returns an empty payload), whereas
    /// PDFKit hard-errors on non-PDF bytes.
    static func detect(from url: URL) -> BookFileType {
        switch url.pathExtension.lowercased() {
        case "pdf": return .pdf
        default:    return .epub
        }
    }
}

/// One book in the user's local library. Anchored on the EPUB file the
/// user picked from disk, NOT on a backend job — the app is primarily a
/// reader, not a conversion manager. The optional `jobId` records the
/// last conversion run for this book, so when the user taps "Save
/// offline" or resumes a partial export we can reattach the audio
/// without re-uploading.
///
/// Persisted as JSON inside `UserDefaults` under key
/// `"library.books.v1"` (see `LibraryStore`).
struct BookEntity: Codable, Identifiable, Hashable {
    /// Stable id derived from the EPUB file content hash; survives
    /// renames and moves of the bookmarked file.
    let id: String

    /// User-visible title. Pulled from EPUB metadata when the book is
    /// imported, with the filename as fallback.
    var title: String

    /// User-visible author. Best-effort from EPUB `dc:creator`.
    var author: String?

    /// Bookmark to the durable app-owned copy of the imported book. On
    /// macOS, legacy rows may initially point to the original external file
    /// and are migrated on their first successful open.
    var bookmark: Data

    /// Filename for display only — never used to access the file. Use
    /// `bookmark` for I/O.
    let displayFilename: String

    /// Imported timestamp.
    let addedAt: Date

    /// Last time the user opened the book (used for "Continue reading"
    /// section).
    var lastOpenedAt: Date?

    /// Resume point inside the book — chapter index + position seconds.
    /// Mirrors what `ResumeStore` keeps per (jobId, chapter).
    var lastChapterIndex: Int?
    var lastPositionSeconds: TimeInterval?

    /// Optional cached cover art. PNG/JPEG bytes, kept inline for now;
    /// large covers can be moved to a separate file cache later.
    var coverPNG: Data?

    /// If the user has run a conversion for this book, we keep the
    /// most recent jobId so we can reattach to the audio assets
    /// (downloadUrl[]) without re-uploading.
    var lastJobId: String?

    /// Whether the user opted in to caching the full audiobook on disk
    /// (the "Save offline" affordance). Controls whether the player
    /// uses streaming TTS or the cached MP3s.
    var cachedOffline: Bool = false

    /// Underlying file format. Defaults to `.epub` for backwards
    /// compatibility with rows persisted before the PDF support
    /// landed — the JSON decoder leaves missing fields at their
    /// default value.
    var fileType: BookFileType = .epub

    var tags: [String] = []

    enum LibraryStatus: String, Hashable, Codable {
        /// EPUB on disk, no audio yet — opening it streams TTS.
        case textOnly
        /// User pressed "Save offline" and a conversion is in flight.
        case caching
        /// Full audio cached locally — playback is offline.
        case offlineReady
    }

    var status: LibraryStatus {
        if cachedOffline { return .offlineReady }
        if lastJobId != nil { return .caching }
        return .textOnly
    }

    var resolvedTitle: String {
        let t = title.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? displayFilename : t
    }

    // MARK: - Codable

    /// Custom decoder so old persisted entries (no `fileType` key)
    /// decode without crashing. The synthesised `Codable` impl treats
    /// every property as required — even with a default value — so we
    /// fall back to `.epub` ourselves and to the path-extension hint
    /// when the legacy entry has a `.pdf` `displayFilename`.
    enum CodingKeys: String, CodingKey {
        case id, title, author, bookmark, displayFilename, addedAt,
             lastOpenedAt, lastChapterIndex, lastPositionSeconds,
             coverPNG, lastJobId, cachedOffline, fileType, tags
    }

    init(
        id: String,
        title: String,
        author: String? = nil,
        bookmark: Data,
        displayFilename: String,
        addedAt: Date,
        lastOpenedAt: Date? = nil,
        lastChapterIndex: Int? = nil,
        lastPositionSeconds: TimeInterval? = nil,
        coverPNG: Data? = nil,
        lastJobId: String? = nil,
        cachedOffline: Bool = false,
        fileType: BookFileType = .epub,
        tags: [String] = []
    ) {
        self.id = id
        self.title = title
        self.author = author
        self.bookmark = bookmark
        self.displayFilename = displayFilename
        self.addedAt = addedAt
        self.lastOpenedAt = lastOpenedAt
        self.lastChapterIndex = lastChapterIndex
        self.lastPositionSeconds = lastPositionSeconds
        self.coverPNG = coverPNG
        self.lastJobId = lastJobId
        self.cachedOffline = cachedOffline
        self.fileType = fileType
        self.tags = tags
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(String.self, forKey: .id)
        self.title = try c.decode(String.self, forKey: .title)
        self.author = try c.decodeIfPresent(String.self, forKey: .author)
        self.bookmark = try c.decode(Data.self, forKey: .bookmark)
        self.displayFilename = try c.decode(String.self, forKey: .displayFilename)
        self.addedAt = try c.decode(Date.self, forKey: .addedAt)
        self.lastOpenedAt = try c.decodeIfPresent(Date.self, forKey: .lastOpenedAt)
        self.lastChapterIndex = try c.decodeIfPresent(Int.self, forKey: .lastChapterIndex)
        self.lastPositionSeconds = try c.decodeIfPresent(TimeInterval.self, forKey: .lastPositionSeconds)
        self.coverPNG = try c.decodeIfPresent(Data.self, forKey: .coverPNG)
        self.lastJobId = try c.decodeIfPresent(String.self, forKey: .lastJobId)
        self.cachedOffline = (try c.decodeIfPresent(Bool.self, forKey: .cachedOffline)) ?? false
        if let raw = try c.decodeIfPresent(BookFileType.self, forKey: .fileType) {
            self.fileType = raw
        } else {
            // Legacy entry — fall back to the filename hint so a
            // re-imported `.pdf` doesn't get stuck reading as EPUB.
            self.fileType = displayFilename.lowercased().hasSuffix(".pdf") ? .pdf : .epub
        }
        self.tags = (try c.decodeIfPresent([String].self, forKey: .tags)) ?? []
    }
}
